tensorflow_quantum/datasets/spin_system_test.py

```diff
@@ -0,0 +1,72 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Test the spin system dataset"""
+import tensorflow as tf
+import numpy as np
+import cirq
+from tensorflow_quantum.datasets import spin_system
+
+
+class SpinSystemDataTest(tf.test.TestCase):
+    """Small test to make sure dataset for ClusterState works."""
+
+    def test_errors(self):
+        """Test that it errors on invalid arguments."""
+        with self.assertRaisesRegex(ValueError,
+                                    expected_regex='Supported number of'):
+            qbs = cirq.GridQubit.rect(3, 1)
+            spin_system.tfi_chain(qbs)
+        with self.assertRaisesRegex(
+                ValueError, expected_regex='Supported boundary conditions'):
+            qbs = cirq.GridQubit.rect(4, 1)
+            spin_system.tfi_chain(qbs, 'open')
+        with self.assertRaisesRegex(TypeError,
+                                    expected_regex='expected str, bytes '):
+            qbs = cirq.GridQubit.rect(4, 1)
+            spin_system.tfi_chain(qbs, data_dir=123)
+        with self.assertRaisesRegex(TypeError,
+                                    expected_regex='must be a list of'):
+            spin_system.tfi_chain(['bob'])
+
+        with self.assertRaisesRegex(TypeError,
+                                    expected_regex='must be a one-dimensional'):
+            qbs = cirq.GridQubit.rect(4, 1)
+            spin_system.tfi_chain([qbs])
+
+    def test_fidelity(self):
+        """Test that it returns the correct number of circuits."""
+        supported_nspins = [4, 8, 12, 16]
+        for nspins in supported_nspins:
+            qbs = cirq.GridQubit.rect(nspins, 1)
+            circuits, _, _, addinfo = spin_system.tfi_chain(
+                qbs,
+                'closed',
+            )
+            fidelities = []
+            for n in range(len(addinfo)):
+                phi = cirq.Simulator().simulate(circuits[n]).final_state
+                gs = addinfo[n].gs
+                fidelities.append(np.abs(np.conj(np.dot(gs, phi))))
+            assert all([np.isclose(fid, 1.0, rtol=1e-3) for fid in fidelities])
+
+    def test_paulisum(self):
+        """Test that hamiltonian returns a PauliSum"""
+        qbs = cirq.GridQubit.rect(4, 1)
+        pauli_sums = spin_system.tfi_chain(qbs, 'closed')[2]
+        assert all(isinstance(ps, cirq.PauliSum) for ps in pauli_sums)
```

Comment:
Can we do a more involved test where you actually compute something using <psi | H | psi > ?

Reply:
I'm having some trouble using `cirq.unitary()` to construct the Hamilonian. Calling `cirq.unitary()` directly on the `PauliSum` does not work since the `PauliSum` does not have a `_unitary()` method. Calling it on the individual `PauliString` terms in the `PauliSum` works only if the coefficient of the `PauliString` has norm one.
line 326 in pauli_string.py:
```python
def _has_unitary_(self) -> bool:
    return abs(1 - abs(self.coefficient)) < 1e-6
```
so this fails:
```python
qbs = cirq.GridQubit.rect(4, 1)
pauli_sums = tfi_chain(qbs, 'closed')[2]
for ps in pauli_sums:
    for term in ps:
        print(term.coefficient)
        U = cirq.unitary(term)
```
```
(-1+0j)
(-1+0j)
(-1+0j)
(-1+0j)
(-0.2+0j)
Traceback (most recent call last):
  File "/home/rooler/Documents/tfq/tensorflow_quantum/datasets/testing_stuff.py", line 9, in <module>
    U = cirq.unitary(term)
  File "/home/rooler/anaconda3/envs/tfq/lib/python3.7/site-packages/cirq/protocols/unitary_protocol.py", line 151, in unitary
    "besides None or NotImplemented.".format(type(val), val))
TypeError: cirq.unitary failed. Value doesn't have a (non-parameterized) unitary effect.

type: <class 'cirq.ops.pauli_string.PauliString'>
value: ((-0.2+0j)*cirq.X(cirq.GridQubit(0, 0)))

The value failed to satisfy any of the following criteria:
- A `_unitary_(self)` method that returned a value besides None or NotImplemented.
- A `_decompose_(self)` method that returned a list of unitary operations.
- An `_apply_unitary_(self, args) method that returned a value besides None or NotImplemented.
```
Also, `cirq.unitary()` returns only local unitaries, not the unitary with respect to the full Hilbert space. So the ZZ terms here are just 4x4 matrices instead of 2^4 x 2^4 (for 4 qubits). This would require you to tensor the individual unitaries based on the qubit number yourself.

Reply:
There is an open issue in Cirq for adding a `.matrix()` function to PauliSum objects. You can do expectation much more compactly with something like:
```python
my_paulisum.expectation_from_wavefunction(numpy_state_vector)
```